lib/wise_homex/api_definition.ex

```diff
@@ -0,0 +1,61 @@
+defmodule WiseHomex.ApiDefinition do
+  @moduledoc """
+  The definition of the Wise Home API
+  """
+
+  defmodule APIResource do
+    @moduledoc """
+    A definition of an api resource
+    """
+
+    @type t :: %__MODULE__{
+            name_singular: String.t(),
+            name_plural: String.t(),
+            path: String.t(),
+            type: String.t()
+          }
+
+    defstruct [:name_singular, :name_plural, :path, :type]
+  end
+
+  @type endpoint_name :: atom()
+
+  @type endpoint_definition :: %{endpoints: [atom], path: String.t(), type: String.t()}
+
+  @type api_definition :: %{endpoint_name => endpoint_definition}
+
+  @doc """
+  The definition of the Wise Home API
+  """
+  @spec api_definition() :: api_definition
+  def api_definition() do
+    %{
+      expense: %{
+        endpoints: [:index, :show, :create, :update, :delete],
+        path: "/expenses",
+        type: "expenses"
+      }
+    }
+  end
+
+  @doc """
+  Transforms the api definition into an APIResource struct representing every function that should be created
+  """
+  @spec to_api_resources(api_definition()) :: [APIResource.t()]
+  def to_api_resources(api_definition) do
+    api_definition
+    |> Enum.flat_map(fn {resource_name, %{endpoints: endpoints, path: path, type: type}} ->
+      endpoints
+      |> Enum.map(fn verb ->
+        api_resource = %APIResource{
+          name_singular: resource_name |> Atom.to_string(),
+          name_plural: type,
```

Comment:
This should snake_case the type (since function names are snake_cased and types are kebab-cased). Or we could just wait with this until we get a type that includes a hyphen.

Reply:
Good point and good catch. Fixed in https://github.com/wise-home/wise_homex/pull/107/commits/46309d70582102f32b563b734cf32a216ec15103

Reply:
Updated in https://github.com/wise-home/wise_homex/pull/107/commits/321d3f49cfeeb9f2852545d49f16123033bef38e with correct function name :-)